src/pip/_internal/resolution/resolvelib/resolver.py

```diff
@@ -147,6 +152,29 @@ def resolve(self, root_reqs, check_supported_wheels):
                 ireq.should_reinstall = True
             elif dist_is_editable(installed_dist) != candidate.is_editable:
                 ireq.should_reinstall = True
+            elif candidate.source_link.is_file:
+                if candidate.source_link.is_wheel:
+                    logger.info(
+                        "%s is already installed with the same version as the "
+                        "provided wheel. Use --force-reinstall to force an "
+                        "installation of the wheel.",
+                        ireq.name,
+                    )
+                    continue
+                if is_archive_file(candidate.source_link.file_path):
```

Comment:
`is_archive_file()` does not always mean sdist since something like `foo.zip` would also return True. I think those should always be re-installed since they are as unreliable as directories.

Reply:
Addressed by special casing `.zip` (there really aren't any good alternatives than directly checking, since `ZIP_EXTENSIONS` includes wheels, which is... meh.